### PR TITLE
Alibaba ram user access keys rotation plugin with spec file

### PR DESCRIPTION
**Pull request by umermehmood**

### CODE REVIEW

File: collectors/alibaba/collector.js [hunk 1]
```diff
@@ -171,6 +171,14 @@ var postcalls = [
...
                 resultKey: 'UserName',
                 apiVersion: '2015-05-01'
+            },
+            ListAccessKeys: {
+                reliesOnService: 'ram',
+                reliesOnCall: 'ListUsers',
+                filterKey: ['UserName'],
+                filterValue: ['UserName'],
+                resultFilter: 'User',
+                apiVersion: '2015-05-01'
```
Comment (AkhtarAmir) [inline collectors/alibaba/collector.js:181]:
```suggestion
            },
            ListAccessKeys: {
                reliesOnService: 'ram',
                reliesOnCall: 'ListUsers',
                filterKey: ['UserName'],
                filterValue: ['UserName'],
                apiVersion: '2015-05-01'
```

File: plugins/alibaba/ram/accessKeysRotation.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            helpers.addResult(results, 0, 'No RAM users found', region);
+            return callback(null, results, source);
+        }
+
+        for (var user of listUsers.data) {
+            if (!user.UserName) continue;
+
+            var getAccessKey = helpers.addSource(cache, source,
+                ['ram', 'ListAccessKeys', region, user.UserName]);
+
```
Comment (AkhtarAmir) [inline plugins/alibaba/ram/accessKeysRotation.js:39]: Add condition to check UNKNOWN case

File: plugins/alibaba/ram/accessKeysRotation.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        for (var user of listUsers.data) {
+            if (!user.UserName) continue;
+
+            var getAccessKey = helpers.addSource(cache, source,
+                ['ram', 'ListAccessKeys', region, user.UserName]);
+            
+            var resource = helpers.createArn('ram', accountId, 'user', user.UserName);
+            var accessKey = getAccessKey.data.AccessKeys.AccessKey.find(key => key.Status && key.Status == 'Active');
```
Comment (AkhtarAmir) [inline plugins/alibaba/ram/accessKeysRotation.js:41]:
```suggestion
            let accessKey;
            if (getAccessKey.data.AccessKeys && getAccessKey.data.AccessKeys.AccessKey && getAccessKey.data.AccessKeys.AccessKey.length) {
                  accessKey = getAccessKey.data.AccessKeys.AccessKey.find(key => key.Status && key.Status == 'Active');
            }
```

File: plugins/alibaba/ram/accessKeysRotation.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                if (diffInDays >= 90) {
+                    helpers.addResult(results, 2,
+                        'RAM user access keys are not rotated every 90 days or less ', region, resource);
+                } else {
+                    helpers.addResult(results, 0,
+                        'RAM user access keys are rotated every 90 days or less', region, resource);
+                }
+            } else {
+                helpers.addResult(results, 0,
+                    'RAM user access keys does not exist ', region, resource);
```
Comment (AkhtarAmir) [inline plugins/alibaba/ram/accessKeysRotation.js:58]:
```suggestion
                    'RAM user does not have any active access keys', region, resource);
```

File: plugins/alibaba/ram/accessKeysRotation.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+            if (accessKey) {
+                let createDate = accessKey.CreateDate;
+                var currentDate = new Date();
+                var createDateFormat = new Date(createDate);
+
+                var diffInDays = helpers.daysBetween(currentDate, createDateFormat);
+                if (diffInDays >= 90) {
+                    helpers.addResult(results, 2,
+                        'RAM user access keys are not rotated every 90 days or less ', region, resource);
```
Comment (AkhtarAmir) [inline plugins/alibaba/ram/accessKeysRotation.js:51]:
```suggestion
                        `RAM user access key is ${diffInDays} days old`, region, resource);
```

File: plugins/alibaba/ram/accessKeysRotation.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                } else {
+                    helpers.addResult(results, 0,
+                        'RAM user access keys are rotated every 90 days or less', region, resource);
```
Comment (AkhtarAmir) [inline plugins/alibaba/ram/accessKeysRotation.js:54]:
```suggestion
                        `RAM user access key is ${diffInDays} days old`, region, resource);
```

File: plugins/alibaba/ram/accessKeysRotation.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        for (var user of listUsers.data) {
+            if (!user.UserName) continue;
+
+            var getAccessKey = helpers.addSource(cache, source,
+                ['ram', 'ListAccessKeys', region, user.UserName]);
+            
+            var resource = helpers.createArn('ram', accountId, 'user', user.UserName);
+            var accessKey = getAccessKey.data.AccessKeys.AccessKey.find(key => key.Status && key.Status == 'Active');
+
+            if (accessKey) {
```
Comment (AkhtarAmir) [inline plugins/alibaba/ram/accessKeysRotation.js:43]: This implementation will only check the first access key. We need to loop all of them and display result for each.

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by giorod3 — APPROVED** [review]

**Review by giorod3 — CHANGES_REQUESTED** [review]
unique results per resource, no duplicates. add setting for rotation length. one result per resource always.

File: plugins/alibaba/ram/accessKeysRotation.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            return callback(null, results, source);
+        }
+
+        for (var user of listUsers.data) {
+            if (!user.UserName) continue;
+
+            var getAccessKey = helpers.addSource(cache, source,
+                ['ram', 'ListAccessKeys', region, user.UserName]);
+            
+            if (getAccessKey.err || !getAccessKey.data) {
```
Comment (giorod3) [inline plugins/alibaba/ram/accessKeysRotation.js:49]:
```suggestion
            if (getAccessKey.err || !getAccessKey.data || !Object.keys(getAccessKey.data).length) {
```

File: plugins/alibaba/ram/accessKeysRotation.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                helpers.addResult(results, 3,
+                    'Unable to query user access keys' + helpers.addError(getAccessKey), region);
+                continue;
+            }
+
+            let resource = helpers.createArn('ram', accountId, 'user', user.UserName);
+            if (getAccessKey.data.AccessKeys && getAccessKey.data.AccessKeys.AccessKey && getAccessKey.data.AccessKeys.AccessKey.length) {
+                let activeKeyFound = false;
+                for (var accessKey of getAccessKey.data.AccessKeys.AccessKey) {
+                    if (accessKey.Status == 'Active') {
```
Comment (giorod3) [inline plugins/alibaba/ram/accessKeysRotation.js:59]:
```suggestion
                    if (accessKey.Status && accessKey.Status == 'Active') {
```

**Review by giorod3 — APPROVED** [review]

**Pull request merged**